Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```
@@ -0,0 +1,196 @@
+            completionHandler(false);
+            return;
+        }
+
+        patterns.add(*pattern);
+    }
+
+    if (!hasPermissions(permissions, patterns)) {
+        completionHandler(false);
+        return;
+    }
+
+    completionHandler(true);;
+}
+
+void WebExtensionContext::request(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool, String)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false, ""_s);
+        return;
+    }
+
+    HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
+    NSString *errorMessage;
+
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"request()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, origins, matchPatterns, @"request()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (hasPermissions(permissions, matchPatterns)) {
+        completionHandler(true, ""_s);
+        return;
+    }
+
+    // FIXME: <https://webkit.org/b/250135> Create and show access alert for requested permissions.
+}
+
+void WebExtensionContext::remove(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool, String)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false, ""_s);
+        return;
+    }
+
+    HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
+    NSString *errorMessage;
+
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"remove()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, origins, matchPatterns, @"remove()", &errorMessage)) {
```

Currently, in Safari we don't report back an error message and just fail silently to match Chrome's behavior. Although, I think passing back an error message in this case might be helpful to developers.

---

Safari does report the error, it is doing it on the web process side via `validatePermissionDetails()`, not in the UI process.